enhancements/machine-api/remediation-history.md

```diff
@@ -0,0 +1,168 @@
+---
+title: remediation-history
+authors:
+  - "@slintes"
+reviewers:
+  - "@beekhof"
+approvers:
+  - "@JoelSpeed"
+  - "@michaelgugino"
+  - "@enxebre"
+creation-date: 2020-12-15
+last-updated: 2020-12-15
+status: implementable
+---
+
+# Remediation history
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Record and show limited remediation history.
+
+## Motivation
+
+With machine health checks administrators can configure automatic remediation of node problems. There is no good
+overview though for when and why nodes were remediated. Recording and showing a history of remediations would allow
+administrators to review when and why machines were remediated, and to discover patterns like e.g. machineA always
+reboots at 2am on Wednesdays.
+
+### Goals
+
+- Record a limited remediation history.
+
+- Show remediation history in the UI.
+
+### Non-Goals
+
+- Record an unlimited history of remediations.
+
+## Proposal
+
+The machine health check holds the configuration for automatic remediation of nodes. We want to extend the status of
+machine health checks to contain the last x remediation events, with information about which node is affected, which
+condition triggered the remediation, and the timestamps of
+- when the triggering condition was detected
+- when remediation was started
+- when the node was succesfully fenced (deleted / powered off)
+- when the node is healthy again.
+
+This information can be displayed in the UI in a table on a machine healthcheck details page.
+
+Example MHC:
+
+```yaml
+apiVersion: machine.openshift.io/v1beta1
+kind: MachineHealthCheck
+metadata:
+  annotations:
+    machine.openshift.io/remediation-strategy: external-baremetal
+  creationTimestamp: "2020-11-24T16:31:22Z"
+spec:
+  maxUnhealthy: 100%
+  nodeStartupTimeout: 60m
+  selector:
+    matchLabels:
+      machine.openshift.io/cluster-api-machine-role: worker
+  unhealthyConditions:
+  - status: "False"
+    timeout: 20s
+    type: Ready
+  - status: Unknown
+    timeout: 20s
+    type: Ready
+status:
+  conditions:
+  - lastTransitionTime: "2020-11-24T17:17:32Z"
+    status: "True"
+    type: RemediationAllowed
+  currentHealthy: 2
+  expectedMachines: 2
+  remediationHistory:
+  - conditionStatus: Unknown
+    conditionType: Ready
+    detected: "2020-11-24T17:33:43Z"
+    fenced: "2020-11-24T17:34:35Z"
+    finished: "2020-11-24T17:35:09Z"
+    remediationType: external
```

Comment:
Not sure what you mean with options here? It is just an indicator that external remediation was used.

Reply:
So what would the value be for this if external wasn't used? This seems like an enum to me, with a set of predefined acceptable values, so I'd like to see a list of the acceptable values defined somewhere (I suspect this list has length 2)

Reply:
Not sure yet. I can think of something like `machineDeletion` for the built in case, and "external" for external remediations at this moment. But a soon as we use the external remediation templates we could use the name of the template. So I thing an enum would not work in the long term. Will add some more info about this.

Reply:
Maybe: builtin, external, and (looking forward to the future) the name of the external template being instantiated

Reply:
that's almost what I already have at https://github.com/openshift/enhancements/pull/567/files#diff-4e98c03501070758bf9d0a5d747bac5fb64f66ff7d3ae0216546dacaaf066d9fR99
How big is the chance that the interna;l / builtin remediation supports multiple strategies? In case we are sure that's only 1 forever, I'm fine with internal or builtin. Otherwise I'd prefer machineDeletion for something more descriptive

Reply:
Ah, I hadn't seen that you'd already addressed this feedback

Reply:
I would lean towards calling it `deletion` or `external`, not sure we need the `machine` prefix as we are talking about MachineHealthChecks and in context I think it makes sense?

Reply:
hm, I see a lot about nodes in the context of MachineHealthChecks as well... so I'd prefer to make it explicit that the machine is deleted by the MHC controller, and not the node (we know this, but maybe not everyone else looking at the MHC status)

Reply:
Ok, given this and your other comment about why machines or nodes, I'm happy to go with `machineDeletion` if you are?